File: 1树的前序遍历中序遍历后序遍历层次遍历和路径/树的前序遍历中序遍历后序遍历层次遍历和路径.py
res = []
def preOrder(root):
    if root:
        res.append(root)
        preOrder(root.left)
        preOrder(root.right)

res = []
def inOrder(root):  
    if root:
        inOrder(root.left)
        res.append(root)
        inOrder(root.right)


res = []
def postOrder(root):
    if root:
        postOrder(root.left)
        postOrder(root.right)
        res.append(root)

res = []
def levelOrder(root):
    temp = [root]
    while(temp):
        node = temp.pop(0)
        res.append(node)
        if node.left:
            temp.append(node.left)
        if node.right:
            temp.append(node.right)


res = []
def findPath(root,path):
    path = path + [root.val]
    if not root.left and not root.right:
        res.append(path)
    if root.left:
        findPath(root.left,path)
    if root.right:
        findPath(root.right,path)

def preOrder(root):
    stack = []
    res = []
    node = root
    while(stack or node):
        if node:
            res.append(node)
            stack.append(node.right)
            node = node.left
        else:
            node = stack.pop(-1)
    return res



def inOrder(root):
    stack = []
    res = []
    node = root
    while(stack or node):
        if node:
            stack.append(node)
            node = node.left
        else:
            node = stack.pop(-1)
            res.append(node)
            node = node.right
    return res

def postOrder(root):
    stack = []
    res = []
    node = root
    while(stack or node):
        if node:
            res.append(node)
            stack.append(node.left)
            node = node.right
        else:
            node = stack.pop(-1)
    return res[::-1]


def findPath(root):
    res = []
    stack = []
    node = root
    stack.append((node,[node.val]))
    while(stack):
        node,path = stack.pop()
        if not node.left and not node.right:
            res.append(path)
        if node.right:
            stack.append((node.right,path + [node.right.val]))
        if node.left:
            stack.append((node.left,path + [node.left.val]))
    return res

File: 1树的前序遍历中序遍历后序遍历层次遍历和路径/树的前序遍历中序遍历后序遍历层次遍历和路径.java
ArrayList<TreeNode> res = new ArrayList<TreeNode>();
public void preOrder(TreeNode root){
    if (root != null){
        res.add(root);
        preOrder(root.left);
        preOrder(root.right);
    }
}


ArrayList<TreeNode> res = new ArrayList<TreeNode>();
public void inOrder(TreeNode root){
    if (root != null){
        inOrder(root.left);
        res.add(root);
        inOrder(root.right);
    }
}


ArrayList<TreeNode> res = new ArrayList<TreeNode>();
public void postOrder(TreeNode root){
    if (root != null){
        postOrder(root.left);
        postOrder(root.right);
        res.add(root);
    }
}


public ArrayList<TreeNode> levelOrder(TreeNode root){
    ArrayList<TreeNode> res = new ArrayList<TreeNode>();
    Queue<TreeNode> temp = new Queue<TreeNode>();
    temp.offer(root);
    while(!temp.empty()){
        TreeNode node = temp.poll();
        res.add(node);
        if (node.left != null){
            temp.offer(node.left);
        }
        if (node.right != null){
            temp.offer(node.right);
        }
    }
    return res;
}



ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
public void findPath(TreeNode root,ArrayList<Integer> path){
    path.add(root.val);
    if (root.left == null && root.right == null){
        res.add(new ArrayList<Integer>(path));
    }
    if (root.left != null){
        findPath(root.left,path);
    }
    if (root.right != null){
        findPath(root.right,path);
    }
    path.remove(path.size() - 1);
}



public ArrayList<TreeNode> preOrder(TreeNode root){
    ArrayList<TreeNode> res = new ArrayList<TreeNode>();
    Stack<TreeNode> temp = new Stack<TreeNode>();
    TreeNode node = root;
    while(!temp.empty() || node != null){
        if (node != null){
            res.add(node);
            temp.push(node.right);
            node = node.left; 
        }else{
            node = temp.pop();
        }
    }
    return res;
}


public ArrayList<TreeNode> inOrder(TreeNode root){
    ArrayList<TreeNode> res = new ArrayList<TreeNode>();
    Stack<TreeNode> temp = new Stack<TreeNode>();
    TreeNode node = root;
    while(!temp.empty() || node != null){
        if (node != null){
            temp.push(node);
            node = node.left; 
        }else{
            node = temp.pop();
            res.add(node);
            node = node.right;
        }
    }
    return res;
}


public ArrayList<TreeNode> postOrder(TreeNode root){
    ArrayList<TreeNode> res = new ArrayList<TreeNode>();
    Stack<TreeNode> temp = new Stack<TreeNode>();
    TreeNode node = root;
    while(!temp.empty() || node != null){
        if (node != null){
            res.add(node);
            temp.push(node.left);
            node = node.right; 
        }else{
            node = temp.pop();
        }
    }
    return Collections.reverse(res);
}

public ArrayList<ArrayList<Integer>> findPath(TreeNode root){
    ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();

    ArrayList<Integer> path = new ArrayList<Integer>();
    Stack<TreeNode> nodeTemp = new Stack<TreeNode>();
    Stack<ArrayList<Integer>> pathTemp = new Stack<ArrayList<Integer>>();
    
    TreeNode node = root;
    path.add(root.val);
    nodeTemp.push(node);
    pathTemp.push(path);
    
    while(!nodeTemp.isEmpty() && !pathTemp.isEmpty()){
        node = nodeTemp.pop();
        path = pathTemp.pop();
        if (node.left == null && node.right == null){
            res.add(new ArrayList<Integer>(path));
        }
        if (node.right != null){
            ArrayList<Integer> p = new ArrayList<Integer>(path);
            p.add(node.right.val);
            nodeTemp.push(node.right);
            pathTemp.push(p);
        }
        if (node.left != null){
            ArrayList<Integer> p = new ArrayList<Integer>(path);
            p.add(node.left.val);
            nodeTemp.push(node.left);
            pathTemp.push(p);
        }
    }
    return res;
}